### PR TITLE
[FLINK-25609][table] Anonymous/inline tables don't require ObjectIdentifier anymore

### PR DESCRIPTION
**Pull request by slinkydeveloper**
## What is the purpose of the change

This PR removes the workaround of generating a random `ObjectIdentifier` for anonymous inline tables, that is:

* Tables defined only with `TableDescriptor`, e.g. `TableEnvironment#from(TableDescriptor)`.
* `DataStream` from/to conversions
* `TableResult#collect`

Now these tables don't go through the `CatalogManager` anymore and they don't have an assigned `ObjectIdentifier` anymore.

Also, for every kind of table (permanent, temporary and anonymous), the table resolved schema, options, and everything inside `ResolvedCatalogTable` is propagated directly through the stack, rather than querying the `Catalog` at a later stage of planning. Among the others, this is a requirement to fix https://issues.apache.org/jira/browse/FLINK-25386.  

## Brief change log

* `DynamicTableFactory.Context#getObjectIdentifier` deprecated and replaced with `DynamicTableFactory.Context#getIdentifier`, which now returns a `Optional<ObjectIdentifier>`.
* Move `TableLookupResult` to the upper level and renamed to `ContextResolvedTable`. Now this POJO is used to propagate `ResolvedCatalogTable`,  `ObjectIdentifier` (if any) and associated `Catalog` (if any) throughout the stack. See the Javadoc for details.
* Renamed `CatalogQueryOperation` and `CatalogSinkModifyOperation` to `SourceQueryOperation` and `SinkModifyOperation` and propagate `ContextResolvedTable`. The renaming is done as the name `Catalog` is confusing and implies that the table can be found inside a catalog, which is not the case anymore for anonymous tables.
* Methods in `Table`, `TableEnvironment` and `StatementSet` creating/accepting anonymous tables now don't generate `ObjectIdentifier` anymore.
* Methods in `Table`, `TableEnvironment` and `StatementSet` using `ObjectIdentifier` will perform a lookup at the call site, rather than later during the planning.
* Now `QueryOperationConverter` manually creates the `RelNode` for anonymous tables, skipping Calcite table resolution
* Every intermediate table representation between the `Operation`s and the final `ExecNode`s now embed `ContextResolvedTable`, rather than `ObjectIdentifier` and `ResolvedCatalogTable`.

This PR contains two breaking changes:

* On master, one could define the pipeline referring to tables yet not available in the catalog, while with this change the invocation of a method like `tEnv.from(String tableName)` will fail if the table has not been loaded in the catalog before. Nevertheless, this shouldn't be a problem for majority of the users and will improve the usability of our APIs, as now they "fail fast".
* Catalog won't store anonymous tables anymore, so people relying on this internal workaround will have to deal with this change defining temporary tables with names. 

Because this is a substantial refactoring, in order to simplify the review, I isolated the most important changes in separate commits. The last commit is a big one, and essentially includes the propagation of the renamings, the propagation of `ContextResolvedTable` through the stack, and little changes to remove the assumption that `ObjectIdentifier` is always non null. 

There is also an hotfix "Add default implementation for `DynamicTableFactory.Context#getEnrichmentOptions` to alleviate breaking change" not related to this PR.

## Verifying this change

All the changes are verified by existing tests. Some tests had been adapted and hardened, see changes to `TableEnvironmentTest`.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: yes
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: yes

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? not applicable

### CODE REVIEW

**Comment by slinkydeveloper** [comment]
@flinkbot run azure

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/plan/schema/TableSourceTable.scala [hunk 1]
```diff
@@ -18,52 +18,52 @@
...
     val flinkContext: FlinkContext,
     val abilitySpecs: Array[SourceAbilitySpec] = Array.empty)
   extends FlinkPreparingTableBase(
     relOptSchema,
     rowType,
-    util.Arrays.asList(
-      tableIdentifier.getCatalogName,
-      tableIdentifier.getDatabaseName,
-      tableIdentifier.getObjectName),
+    toScala(contextResolvedTable.getIdentifier).map(id => util.Arrays.asList(
```
Comment (twalthr) [inline flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/plan/schema/TableSourceTable.scala:63]: improve formatting here or introduce a helper method

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/sinks/TableSinkUtils.scala [hunk 2]
```diff
@@ -56,10 +56,10 @@ object TableSinkUtils {
     * @param partitionKeys The partition keys of this table.
     */
   def validateTableSink(
-      sinkOperation: CatalogSinkModifyOperation,
-      sinkIdentifier: ObjectIdentifier,
-      sink: TableSink[_],
-      partitionKeys: Seq[String]): Unit = {
+                         sinkOperation: SinkModifyOperation,
```
Comment (twalthr) [inline flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/sinks/TableSinkUtils.scala:59]: fix indention

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/operations/ddl/CreateTableASOperation.java [hunk 3]
```diff
@@ -26,25 +26,32 @@
...
 
 /** Operation to describe a CREATE TABLE AS statement. */
 @Internal
 public class CreateTableASOperation implements CreateOperation {
 
     private final CreateTableOperation createTableOperation;
-    private final SinkModifyOperation insertOperation;
+    private final Supplier<SinkModifyOperation> insertOperationFactory;
+
+    private SinkModifyOperation insertOperation;
```
Comment (twalthr) [inline flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/operations/ddl/CreateTableASOperation.java:38]: no mutability in operations, pass the resolved table to `getInsertOperation(...)` instead.

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/catalog/ContextResolvedTable.java [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+    public ContextResolvedTable copy(Map<String, String> newOptions) {
+        if (resolvedTable.getTableKind() == CatalogBaseTable.TableKind.VIEW) {
+            throw new TableException("Cannot copy VIEW with new options.");
+        }
+        return new ContextResolvedTable(
+                objectIdentifier, catalog, ((ResolvedCatalogTable) resolvedTable).copy(newOptions));
+    }
+
+    @Override
+    public String toString() {
```
[the remark this answers is not in the document]
Comment (slinkydeveloper) [reply flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/catalog/ContextResolvedTable.java:160]: I kept it for debugging but i'm not relying on it anymore

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/operations/SinkModifyOperation.java [hunk 5]
```diff
@@ -84,7 +90,7 @@ public QueryOperation getChild() {
     @Override
     public String asSummaryString() {
         Map<String, Object> params = new LinkedHashMap<>();
-        params.put("identifier", tableIdentifier);
+        params.put("identifier", contextResolvedTable);
```
[the remark this answers is not in the document]
Comment (slinkydeveloper) [reply flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/operations/SinkModifyOperation.java]: I reverted and now I print the identifier as before, I think it's better to keep `identifier` as the string key to remain consistent with other operations

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/operations/SourceQueryOperation.java [hunk 6]
```diff
@@ -19,40 +19,45 @@
...
-        return resolvedSchema;
+        return this.contextResolvedTable.getResolvedSchema();
     }
 
     @Override
     public String asSummaryString() {
         Map<String, Object> args = new LinkedHashMap<>();
-        args.put("identifier", tableIdentifier);
-        args.put("fields", resolvedSchema.getColumnNames());
+        args.put("identifier", this.contextResolvedTable.toString());
```
[the remark this answers is not in the document]
Comment (slinkydeveloper) [reply flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/operations/SourceQueryOperation.java]: Same comment as here https://github.com/apache/flink/pull/18349#discussion_r785791932

File: flink-table/flink-table-api-bridge-base/src/main/java/org/apache/flink/table/operations/ExternalQueryOperation.java [hunk 7]
```diff
@@ -93,12 +82,12 @@ public ChangelogMode getChangelogMode() {
     @Override
     public String asSummaryString() {
         final Map<String, Object> args = new LinkedHashMap<>();
-        args.put("identifier", identifier);
+        args.put("identifier", contextResolvedTable);
```
[the remark this answers is not in the document]
Comment (slinkydeveloper) [reply flink-table/flink-table-api-bridge-base/src/main/java/org/apache/flink/table/operations/ExternalQueryOperation.java]: Same comment as here https://github.com/apache/flink/pull/18349#discussion_r785791932

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/QueryOperationConverter.java [hunk 8]
```diff
@@ -340,14 +347,46 @@ private RelNode convertLegacyTableFunction(
...
-            return relBuilder
-                    .scan(
-                            objectIdentifier.getCatalogName(),
-                            objectIdentifier.getDatabaseName(),
-                            objectIdentifier.getObjectName())
-                    .build();
+        public RelNode visit(SourceQueryOperation catalogTable) {
+            Optional<ObjectIdentifier> objectIdentifier =
+                    catalogTable.getContextResolvedTable().getIdentifier();
+            if (objectIdentifier.isPresent()) {
```
[the remark this answers is not in the document]
Comment (slinkydeveloper) [reply flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/QueryOperationConverter.java]: I think we need it for statistics, see `DatabaseCalciteSchema`

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/connectors/DynamicSourceUtils.java [hunk 9]
```diff
@@ -80,28 +79,28 @@ public static RelNode convertDataStreamToRel(
...
             ChangelogMode changelogMode) {
-        final CatalogTable unresolvedTable = new InlineCatalogTable(schema);
-        final ResolvedCatalogTable catalogTable = new ResolvedCatalogTable(unresolvedTable, schema);
         final DynamicTableSource tableSource =
                 new ExternalDynamicSource<>(
-                        identifier, dataStream, physicalDataType, isTopLevelRecord, changelogMode);
+                        dataStream, physicalDataType, isTopLevelRecord, changelogMode);
         final FlinkStatistic statistic =
                 FlinkStatistic.builder()
                         // this is a temporary solution, FLINK-15123 will resolve this
```
[the remark this answers is not in the document]
Comment (slinkydeveloper) [reply flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/connectors/DynamicSourceUtils.java:92]: WDYM? I see this combination of the FlinkStatistic builder usage only once

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/operations/ddl/CreateTableASOperation.java (hunk 3 above)
Comment (slinkydeveloper) [reply flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/operations/ddl/CreateTableASOperation.java:38]: I think I can't, check `asSummaryString`.

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/connectors/DynamicSourceUtils.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (twalthr) [reply flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/connectors/DynamicSourceUtils.java:96]: You added another one below. This comment exists 3 times in the code base now.

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/catalog/ContextResolvedTable.java [hunk 10]
```diff
@@ -0,0 +1,190 @@
...
+        return (T) resolvedTable.getOrigin();
+    }
+
+    /**
+     * Copy the {@link ContextResolvedTable}, replacing the underlying {@link CatalogTable} options.
+     */
+    public ContextResolvedTable copy(Map<String, String> newOptions) {
+        if (resolvedTable.getTableKind() == CatalogBaseTable.TableKind.VIEW) {
+            throw new ValidationException(
+                    String.format("The view '%s' cannot be enriched with new options.", this));
```
Comment (twalthr) [inline flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/catalog/ContextResolvedTable.java:150]: "View '%s' cannot be enriched with new options. Hints can only be applied to tables."

**Review by twalthr — CHANGES_REQUESTED** [review]
Thanks for the update @slinkydeveloper. The PR is in a much better shape now and almost ready to be merged. I added my last feedback. I merged already the hotfix commits. Please rebase, to the latest master.

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/factories/FactoryUtil.java [hunk 11]
```diff
@@ -157,7 +157,7 @@ public static DynamicTableSource createDynamicTableSource(
         } catch (Throwable t) {
             throw new ValidationException(
                     String.format(
-                            "Unable to create a source for reading table '%s'.\n\n"
+                            "Unable to create a source for reading the table '%s'.\n\n"
```
[the remark this answers is not in the document]
Comment (slinkydeveloper) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/factories/FactoryUtil.java:160]: What did you meant with this comment then? https://github.com/apache/flink/pull/18349#discussion_r784698670

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/catalog/ContextResolvedTable.java (hunk 10 above)
Comment (slinkydeveloper) [reply flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/catalog/ContextResolvedTable.java:150]: Added this specific error about hints at the call site.

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/StatementSet.java [hunk 12]
```diff
@@ -62,10 +62,10 @@
      * written to a table (backed by a {@link DynamicTableSink}) expressed via the given {@link
      * TableDescriptor}.
      *
-     * <p>The given {@link TableDescriptor descriptor} is registered as an inline (i.e. anonymous)
-     * temporary catalog table (see {@link TableEnvironment#createTemporaryTable(String,
-     * TableDescriptor)}. Then a statement is added to the statement set that inserts the {@link
-     * Table} object's pipeline into that temporary table.
+     * <p>The given {@link TableDescriptor descriptor} won't be registered in the catalog, but it
+     * will be propagated directly in the operation tree, adding a statement to the statement set
```
[the remark this answers is not in the document]
Comment (slinkydeveloper) [reply flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/StatementSet.java:66]: Reworded, check now

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/operations/ddl/CreateTableASOperation.java (hunk 3 above)
Comment (slinkydeveloper) [reply flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/operations/ddl/CreateTableASOperation.java:38]: I've modified a bit the structure of `CreateTableASOperation` to make the summary string senseful (and similar to `SinkModifyOperation`) and remove that internal mutability. check now

**Review by twalthr — APPROVED** [review]
LGTM

**Pull request closed without merging**